### ISSUE TITLE
Document thread-safety

### ISSUE THREAD

**Issue by qw3ry** [auth0/java-jwt#419]
I would like to know whether or not the classes in this library are thread-safe.

I know there are already two closed issues asking for the thread safety of `JWTVerifier`.
I want to use an `Algorithm` with multiple threads for signing. Is this class threadsafe as well?

I recommend to document the thread-safety in the readme, or at least in the class / method javadocs.

**Comment by lbalmaceda** [comment]
Hi @qw3ry. Thanks for raising this important topic for visibility. 
We agree the thread-safety should be documented carefully along the Javadoc and/or the readme. In the meantime until that analysis and documentation is done, here are some initial thoughts:
- The `Algorithm` class and subclasses are using class variables with the final modifier, and are also not mutating any received parameter. To my understanding, that alone helps a lot to reduce thread issues. 
- The `CryptoHelper` class makes use of `java.security` and `javax.crypto` classes. Nothing fancy, but also nothing unknown. There's no singleton there, every Algorithm instance would have its own helper.
The only "outsiders" could be:
a) the "apache-commons" dependency, used for b64 dec/enc operations. The class is thread-safe according to [their docs](https://commons.apache.org/proper/commons-codec/apidocs/org/apache/commons/codec/binary/Base64.html).
b) the `KeyProvider` implementation of the asymmetric algorithms. What code goes in there is up to the developer. We only invoke that callback in order to get the public key given an ID.
And if anyone would like to contribute to this effort, we'd be happy to collaborate/review any PRs around this effort.

**Comment by qw3ry** [comment]
[It looks like there are some issues with javax.crypto and thread safety](https://stackoverflow.com/a/31898893/2448440)

We chose to be better safe than sorry and create a new Algorithm instance each time, which is a lot better than debugging threading issues later.

**Comment by lbalmaceda** [comment]
@qw3ry This is the result we've reached to: The library is almost completely thread-safe. We've pointed out the things we intend to change to take it to the next level ⚡ 

* There are a few references being kept as received instead of keeping a copy of them. Especially when dealing with arrays, lists, or maps. This can easily be fixed. Classes `PayloadDeserializer`, and `HSAlgorithm`
* The `Date` instance obtained from the clock is mutated. This supposes no problem for the default scenario that ships with this library, as the clock is meant to be customized for test purposes only. We will fix this by cloning the received value before mutating it. Classes `JWTVerifier`
* `Mac` and `Signature` classes (what you pointed out) are used correctly. The "getInstance" method returns a **new instance** of them ready to be used. In addition, methods in `CryptoUtil` use the instance and return their processed result; they never return the Mac or Signature instance. So rest assured, this is thread-safe.

We will make these changes in the coming days. In the meantime, we're interested in hearing your feedback.
